### PR TITLE
Add detection for P7S-Signature

### LINKED ISSUE #123: Add support for p7s file format
1) Specify the MIME type and extension for which to add support
application/pkcs7-signature, .p7s
2) Share an example file
...
3) Optionally, add a reference to the specification of the file format.
https://tools.ietf.org/html/rfc8551
https://github.com/digipres/digipres.github.io/blob/master/_sources/registries/tika/tika-mimetypes.xml#L524

### PR DESCRIPTION
**Pull request by localleon**
This is a PR for #123 and adds the P7S File Format

Closes #123

### CODE REVIEW

**Comment by localleon** [comment]
This PR implements the detection of DEM and PEM P7S Files. 

```
> $ go test .                                                                                                                                                                                        [±p7s ●●]
ok      github.com/gabriel-vasile/mimetype      0.123s                                              
```
Tests complete. But i can't finde out how to run the benchmarks. 

Ready to merge. Hopefully this helps :-)

**Comment by localleon** [comment]
Ready to merge @gabriel-vasile  :-)

Please add the hacktoberfest-accepted label to the Pull Request if you decide to merge it.

File: tree.go [hunk 1]
```diff
@@ -6,14 +6,14 @@ import "github.com/gabriel-vasile/mimetype/internal/matchers"
...
-	sevenZ, zip, pdf, ole, ps, psd, ogg, png, jpg, jp2, jpx, jpm, gif, webp,
+	sevenZ, zip, pdf, ole, ps, psd, p7s, ogg, png, jpg, jp2, jpx, jpm, gif, webp,
 	exe, elf, ar, tar, xar, bz2, fits, tiff, bmp, ico, mp3, flac, midi, ape,
 	musePack, amr, wav, aiff, au, mpeg, quickTime, mqv, mp4, webM, threeGP,
 	threeG2, avi, flv, mkv, asf, aac, voc, aMp4, m4a, m4v, rmvb, utf32le, utf32be,
 	utf16le, utf16be, gzip, class, swf, crx, ttf, woff, woff2, otf, eot, wasm,
 	shx, dbf, dcm, rar, djvu, mobi, lit, bpg, sqlite3, dwg, nes, macho, qcp,
 	icns, heic, heicSeq, heif, heifSeq, mrc, mdb, accdb, zstd, cab, utf8,
-	rpm, xz, lzip, torrent, cpio, tzif,
+	rpm, xz, lzip, torrent, cpio, tzif, p7s,
```
Comment (gabriel-vasile) [inline tree.go:16]: This is not needed, p7s is already registered a few lines above.

File: internal/matchers/signature.go [hunk 2]
```diff
@@ -21,6 +21,30 @@ type (
...
+	if len(in) < 20 {
+		return false
+	}
+	// Magic Bytes for the signedData ASN.1 encoding
+	startHeader := [][]byte{{0x30, 0x80}, {0x30, 0x81}, {0x30, 0x82}, {0x30, 0x83}, {0x30, 0x84}}
+	signedDataMatch := []byte{0x26, 0x06, 0x09, 0x2A, 0x86, 0x48, 0x86, 0xF7, 0x0D, 0x01, 0x07}
+	// Check if Header is correct. There are multiple valid headers
+	for _, match := range startHeader {
+		if bytes.Contains(in[0:2], match) { // If first bytes match -> check for ASN.1 Object Type
+			if bytes.Contains(in[2:20], signedDataMatch) { // only check first part of the ASN.1 Message for OBJECT: signedData
```
Comment (gabriel-vasile) [inline internal/matchers/signature.go:40]: The check for DER described in [tika](https://github.com/apache/tika/blob/75c2ff5686a70c0fb15c4b52534c1be09669af1b/tika-core/src/main/resources/org/apache/tika/mime/tika-mimetypes.xml#L727) is a little different.
When the first two bytes are 0x3080, the 0x06092a864886f70d0107FFa0 bytes are searched for at index 2 in the input.
But if the first two bytes are 0x3081,   the 0x06092a864886f70d0107FFa0 bytes are searched for at offset 3 in the input.

Your code is always checking in[2:20], but it should check 
 - in[3:21] for {0x30, 0x81},
 - in[4:22] for {0x30, 0x82}, and so on

File: tree.go [hunk 3]
```diff
@@ -204,4 +204,6 @@ var (
 	torrent = newMIME("application/x-bittorrent", ".torrent", matchers.Torrent)
 	cpio    = newMIME("application/x-cpio", ".cpio", matchers.Cpio)
 	tzif    = newMIME("application/tzif", "", matchers.TzIf)
+	p7s     = newMIME("application/pkcs7-signature", ".p7s", matchers.P7s).
+		alias(".pem", ".der")
```
Comment (gabriel-vasile) [inline tree.go:208]: `alias` is used to set MIME type aliases, not file extensions.  `application/pkcs7-signature` has no aliases so you can remove this line.

**Review by gabriel-vasile — CHANGES_REQUESTED** [review]
Hi!
The code looks good for the most part, but there are some issues to fix before I merge PR.

File: internal/matchers/signature.go (hunk 2 above)
Comment (localleon) [reply internal/matchers/signature.go:40]: Implementing `i` as a counter value for the offset in the loop should be sufficient

File: tree.go (hunk 1 above)
Comment (localleon) [reply tree.go]: Removed :-)

File: internal/matchers/signature.go [hunk 4]
```diff
@@ -21,6 +21,30 @@ type (
 	}
 )
 
+// P7s matches an .p7s signature File (PEM,Base64)
+func P7s(in []byte) bool {
+	// Check for PEM Encoding
+	if bytes.Contains(in, []byte("BEGIN PKCS7")) {
```
[the remark this answers is not in the document]
Comment (localleon) [reply internal/matchers/signature.go]: Done

**Comment by localleon** [comment]
Implemented the requested changes.

**Comment by localleon** [comment]
@gabriel-vasile Do you need any more changes for this PR to be merged?

**Comment by gabriel-vasile** [comment]
Sorry, I did not have the time to look at it. I will do it this evening. --Sent from my Android phone with mail.com Mail. Please excuse my brevity.On 13/10/2020, 10:29 localleon <notifications@github.com> wrote:

   
  @gabriel-vasile Do you need any more changes for this PR to be merged?

**Review by gabriel-vasile — APPROVED** [review]

**Pull request merged**